python/HeaderService/hslib_salobj.py

```diff
@@ -60,28 +60,29 @@ def __init__(self, **keys):
         # Define the callbacks for start/end
         self.define_evt_callbacks()
 
-        # Initialize the timeout task
-        self.end_evt_timeout_task = salobj.make_done_future()
-
     async def close_tasks(self):
         """Close tasks on super and evt timeout"""
         await super().close_tasks()
-        self.end_evt_timeout_task.cancel()
+        # Loop over all task and cancel all of them
+        for imageName in self.end_evt_timeout_task:
```

Comment:
You'll need a lock around this code, since you're messing with the end_evt_timeout_task list.

Reply:
If the code does not use threads then I think it is not necessary to use locks here because the for loop never releases the event loop.

Reply:
Felipe and I were discussing this in his office while you were making this comment, so that was good timing.  

The hesitation I had in requiring the lock was because of the end_event_timeout() method and the close_tasks() method  since they relinquish the event loop.   if they both run, it's possible that the end_event_timeout method would get called, call clean(), which deletes the end_evt_timeout_task entry.  When the close_tasks() method continues after it's wait, the call to self.end_evt_timeout_task[imageName] would fail.  That's why I thought a lock was needed there.

We had been talking for the more global lock that protects the other data structures;  those probably don't need to be protected, because they never release the event loop.

Reply:
Once the code gets past `await super().close_tasks() it does not relenquish the event loop again. (Neither `self.end_evt_timeout_task.cancel()` nor the for loop that follows releases the event loop). So I think a lock should not be needed and you run the risk of a deadlock if you use the lock *because* the code does not give up the event loop, so the other code that has the lock cannot give it up.

Resource locks are rarely needed in cooperative multitasking code. I do find them necessary once in awhile. For instance in code that issues a command and waits for a reply on a TCP/IP port I will lock that code, so that another task cannot send a command while the first one is waiting for its reply.

Reply:
Thanks @spr and @r-owen for the comments and the time to check the code.